kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -395,36 +397,101 @@ private static void validateSchemaEvolution(
     SchemaChanges schemaChanges = computeSchemaChangesById(currentFieldsById, updatedFieldsById);
-
-    // Validate physical name consistency on the updated fields
     validatePhysicalNameConsistency(schemaChanges.updatedFields());
+    validateUpdatedSchemaCompatibility(schemaChanges);
+    // ToDo Potentially validate IcebergCompatV2 nested IDs
+  }
+
+  /**
+   * Verifies that no non-nullable fields are added, no existing field nullability is tightened and
+   * no invalid type changes are performed
+   *
+   * <p>ToDo: Prevent moving fields outside of their containing struct
+   */
+  private static void validateUpdatedSchemaCompatibility(SchemaChanges schemaChanges) {
+    for (StructField addedField : schemaChanges.addedFields()) {
+      if (!addedField.isNullable()) {
+        throw new KernelException(
+            String.format("Cannot add non-nullable field %s", addedField.getName()));
+      }
+    }
 
-    // ToDo Validate IcebergCompatV2 nested IDs
+    for (Tuple2<StructField, StructField> updatedFields : schemaChanges.updatedFields()) {
+      validateNoTypeChange(updatedFields._1, updatedFields._2);
+    }
+  }
+
+  /**
+   * Validate that there was no change in type from existing field from new field, excluding
+   * modified, dropped, or added fields to structs.
+   */
+  private static void validateNoTypeChange(StructField existingField, StructField newField) {
+    if (existingField.isNullable() && !newField.isNullable()) {
+      throw new KernelException(
+          String.format(
+              "Cannot tighten the nullability of existing field %s", existingField.getName()));
+    }
 
-    // ToDo: Validate schema compatibility
+    // Both fields are structs, ensure there's no changes in the individual fields
+    // ToDo: Prevent additions, removals,
+    //  and type updates to struct fields when the struct is a map key
+    if (existingField.getDataType() instanceof StructType
+        && newField.getDataType() instanceof StructType) {
+      StructType existingStruct = (StructType) existingField.getDataType();
+      StructType newStruct = (StructType) newField.getDataType();
+      Map<Integer, StructField> existingStructFieldsById =
+          existingStruct.fields().stream()
+              .collect(Collectors.toMap(ColumnMapping::getColumnId, Function.identity()));
+
+      for (StructField newNestedField : newStruct.fields()) {
+        StructField existingNestedFields =
+            existingStructFieldsById.get(getColumnId(newNestedField));
+        if (existingNestedFields != null) {
+          validateNoTypeChange(existingNestedFields, newNestedField);
+        }
+      }
+    } else if (existingField.getDataType() instanceof MapType
+        && newField.getDataType() instanceof MapType) {
+      MapType existingMapType = (MapType) existingField.getDataType();
+      MapType newMapType = (MapType) newField.getDataType();
+
+      validateNoTypeChange(existingMapType.getKeyField(), newMapType.getKeyField());
```

Comment:
do we allow dropping or adding columns under the fields of map/array columns? For example if the array element is struct and can that struct be updated to add or remove columns?

Reply:
Yes to arrays. Yes to map keys + values, except for IcebergCompatV2 where we should block for map keys because Iceberg doesn't allow it. Technically the spec doesn't specify, but the implementation is restrictive. On my side, I'm taking a look why we don't allow it in Iceberg impl but for now we can probably be restrictive and loosen it later, once determined on the Iceberg side. See https://github.com/delta-io/delta/pull/4310#issuecomment-2764895347 , that's on the follow on.

The tests already validate that (I have a test with Struct with a field that's an array of structs and a map key where it's a struct)

Reply:
I could also rename to `validateSupportedTypeChange`, arguably validateNoTypeChange isn't an accurate name because if I add a column to a struct in an array that is a type change but it is allowed. The name is also flexible for future type promotion cases

Reply:
Have we validated that we allow this in Delta Spark?

Reply:
I think I agree it's better to be more restrictive now and we can always loosen it in a follow-up

Reply:
I tested in Databricks runtime against a Delta lake table, yeah.